ops/_tracing/export.py

```diff
@@ -0,0 +1,198 @@
+# BUFFER_FILE = '.unit-state.db'
+
+
+_exporter: ProxySpanExporter | None = None
+
+
+class ProxySpanExporter(SpanExporter):
+    real_exporter: SpanExporter | None = None
+
+    def __init__(self, buffer_path: str):
+        self.buffer = ops._tracing.buffer.Buffer(buffer_path)
+
+    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
+        """Export a batch of telemetry data.
+
+        Note: to avoid data loops or recursion, this function cannot be instrumented.
+        """
+        logging.debug(f'FIXME export {len(spans)=}')
+        with suppress_juju_log_handler():
+            # Note:
+            # this is called in a helper thread, which is daemonic,
+            # the MainThread will wait at most 10s for this thread.
+            # Margins:
+            # - 1s safety margin
+            # - 1s for buffered data time overhang
+            # - 2s for live data
+            deadline = time.monotonic() + 6
+
+            assert spans  # the BatchSpanProcessor won't call us if there's no data
+            # TODO:  this will change in the JSON experiment
+            data: bytes = trace_encoder.encode_spans(spans).SerializePartialToString()
+            rv = self.buffer.pump(data)
+            logging.debug('FIXME saved')
+            assert rv
+            self.do_export(*rv)
+
+            for _ in range(SENDOUT_FACTOR - 1):
+                if time.monotonic() > deadline:
+                    break
+                if not (rv := self.buffer.pump()):
+                    break
+                self.do_export(*rv)
+
+            return SpanExportResult.SUCCESS
+
+    def do_export(self, buffered_id: int, data: bytes) -> None:
+        """Export buffered data and remove it from the buffer on success."""
+        logging.debug(f'FIXME asked {buffered_id=} {len(data)=}')
+        # TODO:  this will change in the JSON experiment
+        if self.real_exporter and self.real_exporter._export(data).ok:  # type: ignore
+            logging.debug('FIXME removing')
+            self.buffer.remove(buffered_id)
+
+    def shutdown(self) -> None:
+        """Shut down the exporter."""
+        if self.real_exporter:
+            self.real_exporter.shutdown()
+
+    def force_flush(self, timeout_millis: int = 30000) -> bool:
+        """No-op, as the real exporter doesn't buffer."""
+        return True
+
+    def set_real_exporter(self, exporter: SpanExporter) -> None:
+        self.real_exporter = exporter
+
+
+@contextlib.contextmanager
+def suppress_juju_log_handler():
+    handlers = [h for h in logging.root.handlers if isinstance(h, ops.log.JujuLogHandler)]
+    if not handlers:
+        yield
+        return
+
+    juju_log_handler = handlers[0]
+    token = juju_log_handler.drop.set(True)
+    try:
+        yield
+    finally:
+        juju_log_handler.drop.reset(token)
+
+
+def setup_tracing(charm_class_name: str) -> None:
+    global _exporter
+    # FIXME would it be better to pass Juju context explicitly?
```

Comment:
Yes, please. :-)

Reply:
Let's check with @tonyandrewmeyer 

If I were to pass juju context to this function, I should probably also pass it to _Manager too...

And if so, maybe Scenario's Ops would also have to be changed?

I'm not sure about the trade-offs off the top of my head.

Reply:
Sorry, I missed the notification on this earlier.

The `_Manager` already gets the `_JujuContext` (in theory optionally, but that's only for backwards compatibility and I can definitely get on board with saying we drop that). It's passed through in Scenario now that it uses `_Manager` as well.

So we should definitely pass it through. I think we ought to consider reading from `os.environ` for anything Juju related a bad smell, outside of a single `_JujuContext` creation.

Reply:
Sorry for being super slow here; I think I missed the notification for this.

The `_Manager` already has the `_JujuContext`, and now that Scenario uses `_Manager` as well, it also provides it there. We should definitely pass it to this function.

I think in general we should consider any use of `os.environ` to get Juju information, other than one singular `_JujuContext` creation, a code smell (both in ops/Scenario, and in charms).